File: frontend/types/workshop-application.ts
// Mirrors `industry_workshop_applications`
// (database/migrations/056_workshop_applications.sql) and
// backend/app/schemas/workshop_application.py. Keep all three in sync.
//
// Applicant identity: name/institution/department/graduation year/skills
// are resolved server-side via public.workshop_applicant_profiles (056) --
// never a raw student_id shown as the primary identity. Same fallback
// convention as types/application.ts's `applicantDisplayName`.

export const WORKSHOP_APPLICATION_STATUSES = [
  "APPLIED",
  "ACCEPTED",
  "REJECTED",
  "WITHDRAWN",
  "COMPLETED",
] as const;
export type WorkshopApplicationStatus = (typeof WORKSHOP_APPLICATION_STATUSES)[number];

export const WORKSHOP_APPLICATION_STATUS_LABELS: Record<WorkshopApplicationStatus, string> = {
  APPLIED: "Applied",
  ACCEPTED: "Accepted",
  REJECTED: "Rejected",
  WITHDRAWN: "Withdrawn",
  COMPLETED: "Completed",
};

/** The statuses an Industry account can move a Workshop application to.
 * Mirrors backend `IndustrySettableWorkshopStatus`. */
export const INDUSTRY_SETTABLE_WORKSHOP_STATUSES = ["ACCEPTED", "REJECTED", "COMPLETED"] as const;
export type IndustrySettableWorkshopStatus = (typeof INDUSTRY_SETTABLE_WORKSHOP_STATUSES)[number];

/** Valid Industry-driven transitions. Mirrors
 * industry_workshop_application_service._STATUS_TRANSITIONS. */
export const WORKSHOP_APPLICATION_TRANSITIONS: Record<
  WorkshopApplicationStatus,
  IndustrySettableWorkshopStatus[]
> = {
  APPLIED: ["ACCEPTED", "REJECTED"],
  ACCEPTED: ["COMPLETED", "REJECTED"],
  REJECTED: [],
  WITHDRAWN: [],
  COMPLETED: [],
};

export const WORKSHOP_TRANSITION_LABELS: Record<IndustrySettableWorkshopStatus, string> = {
  ACCEPTED: "Accept",
  REJECTED: "Reject",
  COMPLETED: "Mark completed",
};

export interface WorkshopApplicationOpportunity {
  id: string;
  title: string;
  status: string;
}

export interface WorkshopApplication {
  id: string;
  student_id: string;
  student_name?: string | null;
  institution_name?: string | null;
  department?: string | null;
  graduation_year?: number | null;
  skills?: string[] | null;
  industry_id: string;
  workshop_id: string;
  status: WorkshopApplicationStatus;
  applied_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  workshop: WorkshopApplicationOpportunity | null;
}

export interface WorkshopApplicationListResponse {
  applications: WorkshopApplication[];
}

/** A short, privacy-safe label for an applicant, built only from the id. */
export function workshopApplicantRef(studentId: string): string {
  return `Applicant ${studentId.slice(0, 8)}`;
}

export function workshopApplicantDisplayName(application: {
  student_id: string;
  student_name?: string | null;
}): string {
  const name = application.student_name?.trim();
  return name ? name : workshopApplicantRef(application.student_id);
}

File: backend/app/schemas/workshop_application.py
"""Pydantic schemas for Workshop applications
(`industry_workshop_applications`, database/migrations/056_workshop_applications.sql).

Field names and constraints match that migration exactly. An application
row is created by a STUDENT registering for a PUBLISHED workshop.
Industry's involvement is read + a single mutable field: `status`.
`student_name`/`institution_name`/`department`/`graduation_year`/`skills`
are resolved server-side through `public.workshop_applicant_profiles`, a
SECURITY DEFINER function scoped to the same ownership predicate as this
table's own RLS SELECT policy -- never a raw student_id shown as identity.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict

WorkshopApplicationStatus = Literal["APPLIED", "ACCEPTED", "REJECTED", "WITHDRAWN", "COMPLETED"]

WORKSHOP_APPLICATION_STATUSES: tuple[str, ...] = (
    "APPLIED",
    "ACCEPTED",
    "REJECTED",
    "WITHDRAWN",
    "COMPLETED",
)

# Values an INDUSTRY account is allowed to set. WITHDRAWN belongs to the
# student; APPLIED is the initial state only.
IndustrySettableWorkshopStatus = Literal["ACCEPTED", "REJECTED", "COMPLETED"]


class WorkshopApplicationOpportunity(BaseModel):
    id: str
    title: str
    status: str


class WorkshopApplicationResponse(BaseModel):
    id: str
    student_id: str
    student_name: str | None = None
    institution_name: str | None = None
    department: str | None = None
    graduation_year: int | None = None
    skills: list[str] | None = None
    industry_id: str
    workshop_id: str
    status: str
    applied_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    workshop: WorkshopApplicationOpportunity | None = None


class WorkshopApplicationListResponse(BaseModel):
    applications: list[WorkshopApplicationResponse]


class WorkshopApplyRequest(BaseModel):
    """POST body for a student registering for a workshop. Empty on
    purpose -- there is no cover note field on this table (unlike
    `applications`); kept as a model (not an empty POST) so the route
    shape matches the rest of the API and can grow a field later without
    breaking callers."""

    model_config = ConfigDict(extra="forbid")


class WorkshopApplicationStatusUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: IndustrySettableWorkshopStatus
